inc/addons.php

```diff
@@ -13,17 +13,565 @@
+        $message .= $test;
+        log_error($message);
+        return $message;
+    }
+
+    // test compliancy
+    if (!addon_test_compliancy($declaration['compliancy'])) {
+        $message .= 'Addon not valid compliancy ('.$addon_id.')';
+        log_error($message);
+        return $message;
+    }
+
+    // set declaration in global
+    $GLOBALS['addons'][$addon_id] = $declaration;
+    $GLOBALS['addons'][$addon_id]['enabled'] = addon_test_enabled($addon_id);
+    $GLOBALS['addons'][$addon_id]['_loaded'] = true; // mark it loaded
+    if ($already_loaded !== true) {
+        $GLOBALS['addons'][$addon_id]['settings'] = addon_get_settings($addon_id, $declaration);
+    }
+
+    // new version of an addon
+    if ($new_version === true) {
+        // not really an error...
+        log_error('[Addon updated] The new version of addon '.$addon_id.' seem\'s valid.');
+        // refresh cache
+        if (addons_db_refresh() !== true) {
+            log_error('[Addon updated] Fail to refresh the Addon db');
+        }
+    }
+
+    return true;
+}
+
+
+// addons -> load
+
+/**
+ * load only enabled addons
+ *
+ * @return int the counter of loaded addons (for dev purpose)
+ */
+function addons_load_enabled()
+{
+    $errors = array();
+
+    foreach (addons_list_enabled() as $addon) {
+        if (($loaded = addon_load($addon)) !== true) {
+            $errors[] = $loaded;
+        }
+    }
+    return (count($errors) === 0) ? true : $errors;
+}
+
+/**
+ * load all addons
+ *
+ * @return int the counter of loaded addons (for dev purpose)
+ */
+function addons_load_all()
+{
+    $i = 0;
+    foreach (addons_list_all() as $addon) {
+        if (addon_load($addon)) {
+            ++$i;
+        }
+    }
+
+    return $i;
+}
+
+
+// addon -> get
+
+/**
+ * get 1 setting, addon must be loaded
+ * if addon not loaded or setting do not exist return null
+ *
+ * return mixed, null : fail, other (bool,string...) can be considered as valid value
+ */
+function addon_get_setting($addon_id, $setting_id)
+{
+    if (!isset($GLOBALS['addons'][$addon_id])
+     || !isset($GLOBALS['addons'][$addon_id]['_loaded'])
+     || !isset($GLOBALS['addons'][$addon_id]['settings'][$setting_id])
+     || !isset($GLOBALS['addons'][$addon_id]['settings'][$setting_id]['value'])
+    ) {
+        return null;
+    }
+    return $GLOBALS['addons'][$addon_id]['settings'][$setting_id]['value'];
+}
+
+/**
+ * get all settings for an addon
+ *
+ * TODO : Need to be more desc
+ */
+function addon_get_settings($addon_id, $declaration = null)
+{
+    if (is_null($declaration) && isset($GLOBALS['addons'][$addon_id]) && isset($GLOBALS['addons'][$addon_id]['_loaded'])) {
+        $declaration = $GLOBALS['addons'][$addon_id];
+    }
+
+    // addon dont have $GLOBALS['addons'][]['setting']
+    if (!isset($declaration['settings']) || is_null($declaration['settings']) || !is_array($declaration['settings'])) {
```

Comment:
(micro optimisation) Je l'ai vu aussi plus haut, remplacer `is_null()` par `=== null` est 3 fois plus rapide (http://php.net/manual/fr/function.is-null.php#84161).